base_layer/p2p/src/initialization.rs

```diff
@@ -107,84 +97,130 @@ where
+    // Create outbound channel
+    let (outbound_tx, outbound_rx) = mpsc::channel(10);
 
-    // Connect inbound comms messages to the inbound pipeline and run it
-    ServicePipeline::new(
-        // Messages coming IN from comms to DHT
-        inbound_rx,
-        // Messages going OUT from DHT to connector (pubsub)
-        ServiceBuilder::new()
-            .layer(dht.inbound_middleware_layer())
-            .service(connector),
+    let dht = DhtBuilder::new(
+        comms.node_identity(),
+        comms.peer_manager(),
+        outbound_tx,
+        comms.shutdown_signal(),
     )
-    .with_shutdown_signal(comms.shutdown_signal())
-    .spawn_with(executor.clone());
+    .local_test()
+    .with_discovery_timeout(discovery_request_timeout)
+    .finish();
 
-    //---------------------------------- Outbound Pipeline --------------------------------------------//
+    let dht_outbound_layer = dht.outbound_middleware_layer();
 
-    // Connect outbound message pipeline to comms, and run it
-    ServicePipeline::new(
-        // Requests coming IN from services to DHT
-        dht.take_outbound_receiver().expect("take outbound receiver only once"),
-        // Messages going OUT from DHT to comms
-        ServiceBuilder::new()
-            .layer(dht.outbound_middleware_layer())
-            .service(SinkService::new(outbound_tx)),
-    )
-    .with_shutdown_signal(comms.shutdown_signal())
-    .spawn_with(executor);
+    let comms = comms
+        .with_messaging_pipeline(
+            pipeline::Builder::new()
+                .outbound_buffer_size(10)
+                .with_outbound_pipeline(outbound_rx, |sink| {
+                    ServiceBuilder::new().layer(dht_outbound_layer).service(sink)
+                })
+                .max_concurrent_inbound_tasks(10)
+                .with_inbound_pipeline(
+                    ServiceBuilder::new()
+                        .layer(dht.inbound_middleware_layer())
+                        .service(connector),
+                )
+                .finish(),
+        )
+        .spawn()
+        .await?;
 
     Ok((comms, dht))
 }
 
 /// Initialize Tari Comms
-///
-/// ## Arguments
-///
-/// inbound_connector - Service to call
-pub fn initialize_comms<TSink>(
-    executor: runtime::Handle,
+pub async fn initialize_comms<TSink>(
+    mut config: CommsConfig,
+    connector: InboundDomainConnector<TSink>,
+) -> Result<(CommsNode, Dht), CommsInitializationError>
+where
+    TSink: Sink<Arc<PeerMessage>> + Unpin + Clone + Send + Sync + 'static,
+    TSink::Error: Error + Send + Sync,
+{
+    let builder = CommsBuilder::new().with_node_identity(config.node_identity.clone());
+
+    // This mem::replace is used to prevent a clone - I would prefer to just have transport_type as a separate param,
+    // but this may be less convenient for upstream users of this function
+    match mem::replace(&mut config.transport_type, TransportType::Null) {
```

Comment:
Why do we need to prevent a clone? I see you added a variant to the enum to prevent it too, a security concern?

Reply:
I'll change it - no reason but to prevent a clone which isn't necessary - the tor private key could be in there but does have clear on drop - so it's a silly optimisation.